File: src/main/kotlin/bookvis/models/Chapter.kt
package bookvis.models

import kotlinx.serialization.Serializable

@Serializable
// Updated to support hierarchy and metadata for chapters
// parent is nullable and not recursive in serialization to avoid cycles
// path is a list of strings representing the full hierarchy
// type is 'chapter', 'part', 'book', or 'volume'
data class Chapter(
    val book: Book,
    val title: String,
    val index: Int,
    val parent: Chapter? = null,
    val level: Int? = null,
    val type: String? = null,
    val partTitle: String? = null,
    val globalIndex: Int? = null,
    val path: List<String>? = null,
)

File: src/models/Chapter.ts
import type { Book } from './Book';
import type { Location } from './Location';

export interface Chapter {
  book: Book;
  id: string;                 // Chapter ID like 'chapter-1', 'part-1', etc.
  title: string;
  index: number;
  // Hierarchical structure support
  parent?: Chapter;           // Reference to parent chapter/part
  level?: number;             // Depth level (0 = top level, 1 = sub-level, etc.)
  type?: 'chapter' | 'part' | 'book' | 'volume'; // Type of this node
  // Metadata for display
  partTitle?: string;         // e.g., "Book One", "Part I"

  path?: string[];            // Full path like ["Book 1", "Part 1", "Chapter 1"]
  
  // Locations mentioned in this chapter
  locations?: Location[];
}
